jax/interpreters/partial_eval.py

```diff
@@ -510,7 +525,8 @@ def fun(ki, ui):  # ki will be a known input in this example
     out_pvals = [PartialVal.unknown(ConcreteArray(6)), PartialVal.unknown(ShapedArray)]
     consts = [3, 6]  # values for `ka` and `kb` constvars
   """
-  with core.new_main(JaxprTrace) as main:
+  current_name_stack = source_info_util.current_name_stack()
+  with core.new_main(JaxprTrace, name_stack=current_name_stack) as main:
```

Comment:
Why is this attached as a payload on the MainTrace here, but done via context manager in other places?

Reply:
Other main traces don't need to "remember" the name stack when they were created. The reason we need this functionality for `partial_eval` is because it does both evaluation and staging, so we need to remember the name stack before the trace started so we stage out the jaxpr, we remove the part of the name stack we use in evaluation.

I suppose this could be done via a context manager / store that maps `MainTrace -> NameStack` but since the payload mechanism already exists, I figure that would do. WDYT?